src/QsCompiler/DataStructures/DataTypes.fs

```diff
@@ -53,61 +53,118 @@ type QsNullable<'T> = // to avoid having to include the F# core in the C# part o
-        match left, right with
-        | Value r1, Value r2 -> Value (fst r1, snd r2)
-        | _ -> Null
+/// A position in a text document.
+type Position = private Position of int * int with
+
+    /// The line number, where line zero is the first line in the document.
+    member this.Line = match this with Position (line, _) -> line
+
+    /// The column number, where column zero is the first column in the line.
+    member this.Column = match this with Position (_, column) -> column
+
+    /// Translates the first position by the amount of the second position. If the resulting position is on the same
+    /// line, the column numbers are added; otherwise, the second position's column number is used.
+    static member (+) (a : Position, b : Position) =
+        let line = a.Line + b.Line
+        let column = if b.Line = 0 then a.Column + b.Column else b.Column
+        Position (line, column)
+
+    /// Returns true if the positions have the same line and column numbers.
+    static member op_Equality (a : Position, b : Position) = a = b
+
+    /// Returns true if the second position occurs after the first position. 
+    static member op_GreaterThan (a : Position, b) = (a :> Position IComparable).CompareTo b > 0
+
+    /// <summary>
+    /// Creates a position.
+    /// </summary>
+    /// <exception cref="ArgumentOutOfRangeException">Thrown if the line or column is negative.</exception>
+    static member Create line column =
+        if line < 0 || column < 0
+        then ArgumentOutOfRangeException "Line and column cannot be negative." |> raise
+        Position (line, column)
+
+    /// The position at line zero and column zero.
+    static member Zero = Position (0, 0)
+
+/// A range between two positions in a text document.
+type Range = private Range of Position * Position with
+
+    /// The start of the range.
+    member this.Start = match this with Range (start, _) -> start
+
+    /// The end of the range.
+    member this.End = match this with Range (_, end') -> end'
+
+    /// Adds the range's start and end positions to the given position.
+    static member (+) (position : Position, range : Range) =
+        Range (position + range.Start, position + range.End)
+
+    /// Returns true if the ranges have the same start and end positions.
+    static member op_Equality (a : Range, b : Range) = a = b
+
+    /// <summary>
+    /// Creates a range.
+    /// </summary>
+    /// <exception cref="ArgumentException">
+    /// Thrown if <paramref name="start"/> occurs after <paramref name="end"/>.
+    /// </exception>
+    static member Create start ``end`` =
+        if start > ``end``
+        then ArgumentException "Range start cannot occur after range end." |> raise
+        Range (start, ``end``)
 
-and QsRangeInfo = 
-    QsNullable<QsPositionInfo * QsPositionInfo>
+    /// The empty range starting and ending at position zero.
+    static member Zero = Range (Position.Zero, Position.Zero)
```

Comment:
nit: Might be useful to have a member `Length` on Range. Could return the number of lines spanned and the number of columns spanned, maybe? Although I suppose columns spanned could be negative when lines spanned is greater than zero.

Reply:
I think the most useful definition of length is the number of characters between the start and end, but I think it would need access to the original string to know that. Range length wasn't needed in any of the code I refactored, so I didn't add it.